spec/acceptance/grafana_datasource_spec.rb

```diff
@@ -0,0 +1,161 @@
+                grafana_url          => 'http://localhost:3000',
+                grafana_user         => 'admin',
+                grafana_password     => 'admin',
+                type                 => 'prometheus',
+                url                  => 'https://prom2.example.com',
+                access_mode          => 'proxy',
+                json_data            => {
+                  'httpMethod'       => 'POST',
+                  'timeInterval'     => '10s',
+                },
+                secure_json_data     => {
+                  'basicAuthPassword' => 'prom_password',
+                },
+                basic_auth_user      => 'prom_user',
+              }
+            PUPPET
+          end
+
+          it 'works with no errors' do
+            apply_manifest_on(default, manifest, catch_failures: true)
+          end
+
+          it 'is idempotent' do
+            pending('secure_json_data is not returned by API')
+            apply_manifest_on(default, manifest, catch_changes: true)
+          end
+        end
+      else
+        context 'with basic auth in legacy field' do
+          it_behaves_like 'an idempotent resource' do
+            let(:manifest) do
+              <<-PUPPET
+              grafana_datasource { 'prometheus2':
+                grafana_url          => 'http://localhost:3000',
+                grafana_user         => 'admin',
+                grafana_password     => 'admin',
+                type                 => 'prometheus',
+                url                  => 'https://prom2.example.com',
+                access_mode          => 'proxy',
+                json_data            => {
+                  'httpMethod'       => 'POST',
+                  'timeInterval'     => '10s',
+                },
+                basic_auth_user      => 'prom_user',
+                basic_auth_password  => 'prom_password',
+              }
+              PUPPET
+            end
+          end
+        end
+      end
+    end
+
+    describe 'influxdb ds' do
+      if Gem::Version.new(grafana_version) > Gem::Version.new('9')
+        context 'with password in secure_json_data' do
+          let(:manifest) do
+            <<-PUPPET
+              grafana_datasource { 'influxdb':
+                grafana_url      => 'http://localhost:3000',
+                grafana_user     => 'admin',
+                grafana_password => 'admin',
+                type             => 'influxdb',
+                url              => 'http://localhost:8086',
+                access_mode      => 'proxy',
+                user             => 'admin',
+                secure_json_data => {
+                  'password' => '1nFlux5ecret',
+                },
+                database         => 'mydb',
+              }
+            PUPPET
+          end
+
+          it 'works with no errors' do
+            apply_manifest_on(default, manifest, catch_failures: true)
+          end
+
+          it 'is idempotent' do
```

Comment:
@alexjfisher now with the existing code, is it idempotent but only works on Grafana 8 and older? And the new code isn't idempotent for security data but works on Grafana 9?

Reply:
`secure_json_data` was *never* idempotent, but before grafana 9 you could use the other non secure fields instead.

Reply:
Removing the property defaults (empty strings) from `password` and `basic_auth_password` means that if you were never setting these, upgrading to grafana 9 won't result in the type becoming non idempotent for no reason.